### PR TITLE
[Merged by Bors] - fix(Tactic/Simps): Prevent `simps` from viewing `field_1` as a prefix of `field`.

### PR DESCRIPTION
**Pull request by erdOne**
Related zulip thread: https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/Custom.20name.20for.20projection.20in.20.60structure.20extends.60

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Tactic/Simps/Basic.lean [hunk 1]
```diff
@@ -474,6 +474,25 @@ def findProjectionIndices (strName projName : Name) : MetaM (List Nat) := do
...
+
+private def splitOnNotNumber (s delim : String) : List String :=
+  (process (s.splitOn delim).reverse "").reverse where
+    process (arr : List String) (tail : String) := match arr with
+      | [] => []
+      | (x :: xs) =>
+        if (x.data.head?.elim false (fun x ↦ 48 ≤ x.val ∧ x.val < 58)) then
+          process xs (tail ++ delim ++ x)
+        else
+          List.cons (x ++ tail) (process xs "")
```
Comment (erdOne) [inline Mathlib/Tactic/Simps/Basic.lean:494]: This function is actually not needed to make the test work. I don't really understand the simps code that much to propose a more precise test that could test this.

**Comment by erdOne** [comment]
!bench

**Review by fpvandoorn — COMMENTED** [review]
LGTM

File: Mathlib/Tactic/Simps/Basic.lean [hunk 2]
```diff
@@ -474,6 +474,25 @@ def findProjectionIndices (strName projName : Name) : MetaM (List Nat) := do
   let allProjs := pathToField ++ [fullProjName]
   return allProjs.map (env.getProjectionFnInfo? · |>.get!.i)
 
+/-- We do not consider `toFoo` to be a prefix to `toFoo_1`, as the latter is often
+a different field with an auto-generated name. -/
+private def dropPrefixIfNotNumber? (s : String) (pre : Substring) : Option Substring := do
+  let ret ← Substring.dropPrefix? s pre
+  ret.toString.data.head?.elim ret
+    (fun x ↦ if 48 ≤ x.val ∧ x.val < 58 then none else return ret)
```
[the remark this answers is not in the document]
Comment (erdOne) [reply Mathlib/Tactic/Simps/Basic.lean:482]: Thanks! I couldn't find them.

File: Mathlib/Tactic/Simps/Basic.lean (hunk 1 above)
Comment (erdOne) [reply Mathlib/Tactic/Simps/Basic.lean:494]: I just realized that previous occasions of `String.splitOn s "_"` were all for tracing purposes and that's why I failed to test them. The tests are sill useful though and I'll add them. Thanks!

**Comment by fpvandoorn** [comment]
bors merge

**Pull request closed without merging**